### ISSUE TITLE
[1.15.2] All items dissapeared after enabling auto pickup

### ISSUE THREAD

**Issue by Tallivm** [Flanks255/simplybackpacks#20]
My uncommon backpack was full of goodies when I enabled auto-pickup (shift+right click and turned on the slider). After that, I picked up some other items and looked inside the backpack. All the items were gone, there were only a couple of stone blocks etc. which were just picked.  

I have other mods as well, there were both vanilla and mod items which disappeared. I have Mouse Tweaks.  

Forge version: 31.1.47
Mod version: 1.15.2-1.4.5

**Comment by Flanks255** [comment]
Is this on a single player world, multiplayer?
Can you reproduce this?
Ive tried in both SP and 2 different servers and so far cant reproduce it myself.

**Comment by Tallivm** [comment]
It was a single player world opened for LAN on my computer.
It was about 5th or 6th time I enabled auto pick-up for this backpack in this world when it was full, and the first time when items disappeared. I will try to reproduce the bug, but looks like it won't happen every time.

**Comment by Cocomere** [comment]
I've had the same thing happen 3-4 times. I have two fully upgraded backpacks I keep on me, though I've had it happen with only one.

The first time, I had two blue backpacks, and I had toggled auto pickup a few times trying to get both to be active at once. (Dumb, I know) When I went to put the items I had picked up away, everything that had been in there before was gone. I assumed this was due to the toggling and just cheated to get my items back.

The second time, I was in the nether (both backpacks now fully upgraded), and had kept auto pickup on without messing with it for a day or so of play. I had raided a good 4-5 nether fortresses, quit for the day, and then came back and immediately went back to the overworld. When I checked my first backpack, nearly everything was gone. Again, I cheated to give myself the most important items again, but I lost a LOT of good loot.

The most recent time happened about ten minutes ago. I was in the overworld, farming granite, and everything was fine. I had had autopickup on for a while, no issues. I went back home, messed around with a crusher setup to get a better hopper, and checked my second backpack without meaning to. When I did, I realized that all of my items I kept in there were gone, replaced by some of the stuff that had been in the hopper.

Is it possible that there are issues with having two backpacks and autopickup? Like Tallivm, this doesn't happen all the time. I don't know that I can reproduce it.

For reference, I'm using the Enigmatica 4 modpack through the twitch launcher. The only modification I've made is to remove angel ring, and this was happening both before and after I did that.

**Comment by Cocomere** [comment]
Okay, update: I quit the game without saving, and when I returned I was exactly where I had been when I exited, but all the items in my second backpack were back. I'm glad, but more confused.

The first time this happened, I saved and exited the normal way, but this did not fix it.

**Comment by Flanks255** [comment]
thats strange, almost sounds like some sort of twisted desync...

**Comment by Cocomere** [comment]
Oh, and the second time, it was ONLY the first backpack that was messed up. The second one was perfectly fine.

**Comment by fchud** [comment]
hi.

lost all contents of a basic backpack.
had some items in it; it was on auto-pickup.

i died to an arrow in a face and when i came back to pickup everything - items that were not in a backpack got auto-picked up into backpack replacing all it's contents.

single player, 1.15.2-forge-31.2.0 and some mods:

- advancedtools-1.15.2-1.0.2.jar
- autoattack-1.0.jar
- BetterThanMending-1.4.0.jar
- BowInfinityFix-1.15.2-rv8.jar
- branders-enhanced-mob-spawners-v1.15.2-1.5.4.jar
- bspawncontrol_1.15.2-1.0.jar
- CraftingTweaks_1.15.1-11.0.0.jar
- Cucumber-1.15.2-3.0.5.jar
- curios-FORGE-1.15.2-2.0.2.2.jar
- dotHack-Weapons-1.15.2-1.3.2.jar
- essentials-1.15.2-2.7.1.jar
- extendedfurnace-2.2.jar
- ferroustry-2.2.jar
- FluxNetworks-1.15.2-5.0.1-2.jar
- forgemod_VoxelMap-1.9.18b_for_1.15.2.jar
- FpsReducer-mc1.15.2-1.15.jar
- GameMenuModOption-1.15.2-1.2.jar
- glasscutter-1.15.2-1.0.1.jar
- industrial-foregoing-1.15.2-2.3.0-6f57752.jar
- infinitemusic-1.1.0.jar
- inventorysorter-1.15.2-17.0.0.jar
- ironchest-1.15.2-10.0.3.jar
- ironfurnaces-1.15.2-2.0.3.jar
- IronJetpacks-1.15.2-3.0.1.jar
- jei-1.15.2-6.0.0.4.jar
- jeiintegration_1.15.2-3.1.0.1.jar
- lithium-mc1.15.2-forge-0.4.6-rc3-mod.jar
- Lollipop-1.15.2-1.0.15.jar
- malilib-forge-1.15.2-0.10.0-dev.20+beta.1.jar
- mcjtylib-1.15-4.1.0-beta.jar
- Mine+and+Slash-1.15.2-3.6.14.jar
- minihud-forge-1.15.2-0.19.0-dev.20200412.215359.jar
- MouseTweaks-2.13-mc1.15.1.jar
- moveboats_1.15.2-1.1.jar
- moveminecarts_1.15.2-1.2.jar
- NaturesCompass-1.15.2-1.8.5.jar
- Neat+1.6-22.jar
- paperbooks_1.15.2-1.0.jar
- Patchouli-1.15.2-1.2-33.jar
- phosphor-forge-mc1.15.2-0.5.2+build.4.jar
- PickleTweaks-1.15.2-4.0.2.1.jar
- Powah-1.15.2-1.1.14.jar
- PrettyBeaches_1.15.1-4.0.0.jar
- refinedstorage-1.8.3.jar
- refinedstorageaddons-0.6.1.jar
- rftoolsbase-1.15-1.1.1-beta.jar
- rftoolspower-1.15-2.1.1-beta.jar
- rftoolsstorage-1.15-1.1.2-beta.jar
- rftoolsutility-1.15-2.1.2-beta.jar
- RingOfAttraction-1.15.2-1.1.31.jar
- SilentGear-1.15.2-1.6.4+90.jar
- SilentLib-1.15.2-4.6.1+49.jar
- SilentMechanisms-1.15.2-0.7.1+51.jar
- simplybackpacks-1.15.2-1.4.5.jar
- SlimeBottles.jar
- StepUp-1.15.2-0.2.1.jar
- SuperMultiDrills-1.15.2-2.0.1+3.jar
- swingthroughgrass-1.15.2-1.4.0.jar
- tesseract-1.0.4-mc1.15.jar
- theoneprobe-1.15-2.0.2.jar
- titanium-1.15.2-2.4.0.jar
- toms_storage-1.0.2.jar
- ToolBelt-1.15.2-1.14.1.jar
- tpsgenerator-1.0.jar
- TrashSlot_1.15.2-11.0.0.jar
- upsizer-mc1.15.2-5.1.jar
- vanilla_steel-1.0.1.jar
- VTweaks-1.15.2-3.1.6.jar
- walljump-forge-1.15.2-1.3.5.jar
- wirelessredstone-1.15.2-0.1.0.jar
- xray-1.15.2-2.2.0.jar
- [Forge]+[1.15.2]+GamemodeOverhaul-1.15.2-1.4.2.2.jar

**Comment by Tallivm** [comment]
Thought I should post my list of mods, too. Listing only those which fchud has as well:

- ironchest-1.15.2-10.0.3.jar
- ironfurnaces-1.15.2-2.0.2.jar
- jei-1.15.2-6.0.0.3.jar
- MouseTweaks-2.13-mc1.15.1.jar
- NaturesCompass-1.15.2-1.8.5.jar
- SilentGear-1.15.2-1.6.0+86.jar
- SilentLib-1.15.2-4.6.0+48.jar

**Comment by fchud** [comment]
just tested:

after you pickup a **_backpack with enabled auto-pickup_**, next thing picked up will clear its entire inventory.
this also applies to moving a backpack _from a chest or other inventories_ to your inventory.

however, if you open a **_backpack with enabled auto-pickup_** after picking up - all things continue working as intended.

**Comment by Cocomere** [comment]
> just tested:
> 
> after you pickup a **_backpack with enabled auto-pickup_**, next thing picked up will clear its entire inventory.
> this also applies to moving a backpack _from a chest or other inventories_ to your inventory.
> 
> however, if you open a **_backpack with enabled auto-pickup_** after picking up - all things continue working as intended.

This could be part of the issue, however in all instances where this has happened to me (And I've had another two or three instances) I haven't put down my backpacks, or even moved them from the quickbar. 

Of the mods others have listed, I think the only ones we have in common are iron chests and JEI. I'm not really sure how either would cause an issue, but then again I'm not a programmer. 

I also realized I'm not positive what update my modpack is for both this mod and the game. I should probably check that once I'm not on my phone.

**Comment by ponarC** [comment]
Confirm issue on simplybackpacks1.15.2-1.4.5 (FTB Omnia)
Reproduce steps like @fchud 
I have few rare backpacks, all have enabled auto-pickup.
When move last backpack in hot bar to earlier place to guarantee that they pickup item first. After first pickup all items in that backpack is lost. But if  i open backpack first after move, then content remain after item pickup.

I inspect code and missing load() in constructor in BackpackItemHandler or BackpackCaps looks suspicious. It's create every time when stack is created, split or moved to new slot. They create new ItemStackHandler with is empty by default and copy ItemStack with contain real backpack content in serialized format. When item is picked-up it inserted to empty ItemStackHandler and save is called for sync with ItemStack serialized content. Yup previous content is lost and backpack pick-up start from scratch ...

But if backpack is opened.. SBContainer is create to present backpack content. They called load on BackpackItemHandler for sync content with ItemStack. Yup previous content is restored (Looks like simple fix for empty backpack after backpack move).

**Comment by davqvist** [comment]
I just released a modpack with your mod in it and people are reporting this bug which is pretty major.
I wonder if it is still being worked on and can the above fix be implemented?

**Comment by gloomyMoron** [comment]
I just had this same bug happen, but my backpack wasn't dropped. What I think happened was that there was enough space for one or two more of a stack with every other slot filled and when it went to pick up the stack that put it over its capacity limit, it emptied the pack. I *don't think* I dropped it, but I was panicking because I was dying of suffocation (I lived and didn't die but my bag was empty).
